src/sentry/middleware/integrations/parsers/base.py

```diff
@@ -0,0 +1,110 @@
+        self.request = request
+        self.match: ResolverMatch = resolve(self.request.path)
+        self.response_handler = response_handler
+        self.error_message = "Integration Request Parsers should only be run on the control silo."
+
+    def _ensure_control_silo(self):
+        if SiloMode.get_current_mode() != SiloMode.CONTROL:
+            logger.error(
+                "integration_control.base.silo_error",
+                extra={"path": self.request.path, "silo": SiloMode.get_current_mode()},
+            )
+            raise SiloLimit.AvailabilityError(self.error_message)
+
+    def get_response_from_control_silo(self):
+        """
+        Used to handle the request directly on the control silo.
+        """
+        self._ensure_control_silo()
+        return self.response_handler(self.request)
+
+    def get_response_from_region_silos(self, regions: Iterable[Region]):
+        """
+        Used to handle the requests on a given list of regions (synchronously).
+        If multiple regions are provided, only the last response is returned to the requestor.
+        """
+        self._ensure_control_silo()
+        region_response = None
+        for region in regions:
+            region_client = RegionSiloClient(region)
+            region_response = region_client.proxy_request(self.request).to_http_response()
+        if region_response is None:
+            logger.error(
+                "integration_control.base.region_proxy_error",
+                extra={
+                    "path": self.request.path,
+                    "regions": [region.name for region in regions],
+                },
+            )
+            return self.response_handler(self.request)
+        return region_response
+
+    def get_response(self):
+        """
+        Used to surface a response as part of the middleware.
+        Should be overwritten by implementation.
+        Default behaviour is handle the response ignoring SiloMode.
+        """
+        return self.response_handler(self.request)
+
+    def get_integration(self) -> Integration | None:
+        """
+        Parse the request to retreive organizations to forward the request to.
+        Should be overwritten by implementation.
+        """
+        return None
+
+    def get_organizations(self, integration: Integration = None) -> Sequence[Organization]:
+        """
+        Use the get_integration() method to identify organizations associated with
+        the integration request.
+        """
+        if not integration:
+            integration = self.get_integration()
+        if not integration:
+            logger.error(
+                "integration_control.base.no_integration",
+                extra={"path": self.request.path},
+            )
+            return []
+        organization_integrations = OrganizationIntegration.objects.filter(
+            integration_id=integration
+        ).select_related("organization")
```

Comment:
this relationship will likely need to be broken as organization won't be in Control Silo.

Reply:
Makes sense, I will use the organization_service instead :+1: